pySDC/projects/DAE/sweepers/SemiImplicitDAE.py

```diff
@@ -0,0 +1,190 @@
+
+    References
+    ----------
+    .. [1] J. Huang, J. Jun, M. L. Minion. Arbitrary order Krylov deferred correction methods for differential algebraic
+       equation. J. Comput. Phys. Vol. 221 No. 2 (2007).
+
+    Note
+    ----
+    The right-hand side of the problem DAE classes using this sweeper has to be exactly implemented in the way, the
+    semi-explicit DAE is defined. Define :math:`\vec{x}=(y, z)^T`, :math:`F(\vec{x})=(f(\vec{x}), g(\vec{x}))`, and the
+    matrix
+
+    .. math::
+        A = \begin{matrix}
+            I & 0 \\
+            0 & 0
+        \end{matrix}
+
+    then, the problem can be reformulated as
+
+    .. math::
+        A\vec{x}' = F(\vec{x}).
+
+    Then, setting :math:`F_{new}(\vec{x}, \vec{x}') = A\vec{x}' - F(\vec{x})` defines a DAE of fully-implicit form
+
+    .. math::
+        0 = F_{new}(\vec{x}, \vec{x}').
+
+    Hence, the method ``eval_f`` of problem DAE classes of semi-explicit form implements the right-hand side in the way of
+    returning :math:`F(\vec{x})`, whereas ``eval_f`` of problem classes of fully-implicit form return the right-hand side
+    :math:`F_{new}(\vec{x}, \vec{x}')`.
+    """
+
+    def __init__(self, params):
+        """Initialization routine"""
+
+        if 'QI' not in params:
+            params['QI'] = 'IE'
+
+        # call parent's initialization routine
+        super().__init__(params)
+
+        msg = f"Quadrature type {self.params.quad_type} is not implemented yet. Use 'RADAU-RIGHT' instead!"
+        if self.coll.left_is_node:
+            raise ParameterError(msg)
+
+        self.QI = self.get_Qdelta_implicit(coll=self.coll, qd_type=self.params.QI)
+
+    def integrate(self):
+        r"""
+        Returns the solution by integrating its gradient (fundamental theorem of calculus) at each collocation node.
+        ``level.f`` stores the gradient of solution ``level.u``.
+
+        Returns
+        -------
+        me : list of lists
+            Integral of the gradient at each collocation node.
+        """
+
+        # get current level and problem description
+        L = self.level
+        P = L.prob
+        M = self.coll.num_nodes
+
+        me = []
+        for m in range(1, M + 1):
+            # new instance of dtype_u, initialize values with 0
+            me.append(P.dtype_u(P.init, val=0.0))
+            for j in range(1, M + 1):
+                me[-1].diff[:] += L.dt * self.coll.Qmat[m, j] * L.f[j].diff[:]
+
+        return me
+
+    def update_nodes(self):
+        r"""
+        Updates the values of solution ``u`` and their gradient stored in ``f``.
+        """
+
+        L = self.level
+        P = L.prob
+
+        # only if the level has been touched before
+        assert L.status.unlocked
+        M = self.coll.num_nodes
+
+        integral = self.integrate()
+        # build the rest of the known solution u_0 + del_t(Q - Q_del)U_k
+        for m in range(1, M + 1):
+            for j in range(1, M + 1):
```

Comment:
this range should only go to $m+1$. All other entries in the preconditioner are zero. And we don't need to multiply many solution size objects by zero and then add that. It's also done like this in the generic implicit sweeper, but I think we should rather change it in both. The rest of this function depends on the lower triangular structure of `QI`, such that this does not reduce generality, if I am not mistaken.